plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -28,41 +28,91 @@
 import com.google.cloud.storage.Storage;
 import com.google.cloud.storage.StorageOptions;
 
+import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.collect.MapBuilder;
 import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
-import org.elasticsearch.env.Environment;
-
 import java.io.IOException;
 import java.net.HttpURLConnection;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.util.Map;
 
+import static java.util.Collections.emptyMap;
+
 public class GoogleCloudStorageService extends AbstractComponent {
 
     /** Clients settings identified by client name. */
-    private final Map<String, GoogleCloudStorageClientSettings> clientsSettings;
+    private volatile Map<String, GoogleCloudStorageClientSettings> clientsSettings = emptyMap();
+    /** Cache of client instances. Client instances are built once for each setting change. */
+    private volatile Map<String, Storage> clientsCache = emptyMap();
+
+    public GoogleCloudStorageService(final Settings settings) {
+        super(settings);
+    }
 
-    public GoogleCloudStorageService(final Environment environment, final Map<String, GoogleCloudStorageClientSettings> clientsSettings) {
-        super(environment.settings());
-        this.clientsSettings = clientsSettings;
+    /**
+     * Updates the client settings and clears the client cache. Subsequent calls to
+     * {@code GoogleCloudStorageService#client} will return new clients constructed
+     * using these passed settings.
+     *
+     * @param clientsSettings the new settings used for building clients for subsequent requests
+     * @return previous settings which have been substituted
+     */
+    public synchronized Map<String, GoogleCloudStorageClientSettings>
```

Comment:
We could have a `refreshAndClearCache(final Settings settings)` method or with a similar name that reflects the fact that settings are refresh and cached clients deleted. 

We could also use `GoogleCloudStorageClientSettings.load()` here that already returns an immutable map.

```
public synchronized refreshAndClearCache(final Settings settings) {
        final Map<String, GoogleCloudStorageClientSettings> prevSettings = this.clientsSettings;
        this.clientsSettings = GoogleCloudStorageClientSettings.load(settings);
        ...
    }
```

What do you think? Unless not *all* the settings are passed when the plugin is reinitialized?

Reply:
> We could have a refreshAndClearCache(final Settings settings) method or with a similar name that reflects the fact that settings are refresh and cached clients deleted.

Agreed.

Again, I prefer to read secure settings in a `GoogleCloudStorageClientSettings` instance and pass that around, not the raw `Settings` object.

Reply:
Makes sense, thanks. Do you think it could deserve a comment?